lib/translator/report_luns.cc

```diff
@@ -0,0 +1,110 @@
+}
+
+}  // namespace
+
+// Section 4.5
+// https://www.nvmexpress.org/wp-content/uploads/NVM-Express-SCSI-Translation-Reference-1_1-Gold.pdf
+StatusCode ReportLunsToNvme(absl::Span<const uint8_t> scsi_cmd,
+                            nvme_defs::GenericQueueEntryCmd& identify_cmd,
+                            uint32_t& alloc_len) {
+  if (scsi_cmd.size() != sizeof(scsi_defs::ReportLunsCommand)) {
+    DebugLog("Malformed ReportLuns command");
+    return StatusCode::kInvalidInput;
+  }
+
+  // Cast scsi_cmd to ReportLunsCommand
+  scsi_defs::ReportLunsCommand rl_cmd;
+  ReadValue(scsi_cmd, rl_cmd);
+  if (rl_cmd.select_report != scsi_defs::SelectReport::kRestrictedMethods &&
+      rl_cmd.select_report != scsi_defs::SelectReport::kWellKnown &&
+      rl_cmd.select_report != scsi_defs::SelectReport::kAllLogical) {
+    DebugLog("Invalid report luns select report %u", rl_cmd.select_report);
+    return StatusCode::kInvalidInput;
+  }
+
+  // Assign allocation length for downstream use
+  alloc_len = rl_cmd.alloc_length;
+
+  // Assign generic params to NVMe command
+  identify_cmd = {};
+  identify_cmd.opc = static_cast<uint8_t>(nvme_defs::AdminOpcode::kIdentify);
+  identify_cmd.cdw[0] = 0x2;  // Set CNS to return namespace ID list
+
+  // Allocate prp & assign to command
+  uint64_t prp_ptr = AllocPages(1);
+  if (prp_ptr == 0) return StatusCode::kFailure;
+  identify_cmd.dptr.prp.prp1 = prp_ptr;
+  return StatusCode::kSuccess;
+}
+
+// Section 6.6
+// https://www.nvmexpress.org/wp-content/uploads/NVM-Express-SCSI-Translation-Reference-1_1-Gold.pdf
+StatusCode ReportLunsToScsi(const nvme_defs::GenericQueueEntryCmd& identify_cmd,
+                            absl::Span<uint8_t> buffer) {
+  if (buffer.size() < sizeof(scsi_defs::ReportLunsParamData)) {
+    DebugLog("Insufficient buffer size");
+    return StatusCode::kFailure;
+  }
+
+  // Get ns_list and lun count from NVMe dptr
+  nvme_defs::IdentifyNamespaceList ns_list;
+  uint8_t* dptr = reinterpret_cast<uint8_t*>(identify_cmd.dptr.prp.prp1);
+  ReadValue(absl::MakeSpan(dptr, sizeof(ns_list)), ns_list);
+  uint32_t lun_count = GetNsListLength(ns_list);
+
+  // Define scsi response structure and determine attributes
+  uint32_t allocated_list_bytes =
+      buffer.size() - sizeof(scsi_defs::ReportLunsParamData);
+  scsi_defs::ReportLunsParamData rlpd = {};
+  rlpd.list_byte_length = lun_count * sizeof(scsi_defs::LunAddress);
+  if (rlpd.list_byte_length > allocated_list_bytes) {
+    rlpd.list_byte_length =
+        allocated_list_bytes -
+        (allocated_list_bytes % sizeof(scsi_defs::LunAddress));
+    lun_count = rlpd.list_byte_length / sizeof(scsi_defs::LunAddress);
+  }
+
+  // Write response to buffer
+  WriteValue(rlpd, buffer);
+  uint32_t lun_offset = sizeof(scsi_defs::ReportLunsParamData);
+  for (uint32_t i = 0; i < lun_count; ++i) {
+    scsi_defs::LunAddress lun = static_cast<scsi_defs::LunAddress>(
+        ns_list.ids[i] - 1);  // LUN must start @ 0 via SAM spec
```

Comment:
Are other parts of the project consistently applying this same transformation?  Do we want a shared helper function to formalize mapping a LUN to an NSID and back?

Reply:
Great point! Currently other parts of the project would have to apply this same transformation, but I think it's reasonable to transform the lun to nsid in the "begin" function, and then downstream code can use it without knowing translation details